superset-websocket/src/index.ts

```diff
@@ -0,0 +1,329 @@
+  error_type: string;
+  extra: ExtraType;
+  level: ErrorLevel;
+  message: string;
+};
+
+type ListenerFunction = (results: StreamResult[]) => void;
+interface EventValue {
+  id: string,
+  channel_id: string,
+  job_id: string,
+  user_id?: string,
+  status: string,
+  errors?: SupersetError[],
+  result_url?: string,
+}
+interface JwtPayload { channel: string }
+interface FetchRangeFromStreamParams { sessionId: string, startId: string, endId: string, listener: ListenerFunction }
+export interface SocketInstance { ws: WebSocket, channel: string, pongTs: number }
+
+interface ChannelValue {
+  sockets: Array<string>,
+}
+
+const environment = process.env.NODE_ENV;
+export const opts = {
+  port: 8080,
+  redis: {
+    port: 6379,
+    host: "127.0.0.1",
+    password: "",
+    db: 0
+  },
+  streamPrefix: "async-events-",
+  jwtSecret: "",
+  jwtCookieName: "async-token",
+  redisStreamReadCount: 100,
+  redisStreamReadBlockMs: 5000,
+  socketResponseTimeoutMs: 60 * 1000,
+  pingSocketsIntervalMs: 20 * 1000,
+  gcChannelsIntervalMs: 120 * 1000,
+}
+
+const startServer = process.argv[2] === 'start';
+const configFile = environment === 'test' ? '../config.test.json' : '../config.json';
+let config = {};
+try {
+  config = require(configFile);
+} catch(err) {
+  console.warn('config.json not found, using defaults');
+}
+
+Object.assign(opts, config);
+
+if(startServer && opts.jwtSecret.length < 32)
+  throw('Please provide a JWT secret at least 32 bytes long')
+
+const redis = new Redis(opts.redis);
+const httpServer = http.createServer();
+export const wss = new WebSocket.Server({ noServer: true, clientTracking: false });
+
+const SOCKET_ACTIVE_STATES = [WebSocket.OPEN, WebSocket.CONNECTING];
+const GLOBAL_EVENT_STREAM_NAME = `${opts.streamPrefix}full`;
+const DEFAULT_STREAM_LAST_ID = '$';
+
+export let channels: Record<string, ChannelValue> = {};
+export let sockets: Record<string, SocketInstance> = {};
+let lastFirehoseId: string = DEFAULT_STREAM_LAST_ID;
+
+
+export const setLastFirehoseId = (id: string): void => {
+  lastFirehoseId = id;
+}
+
+export const trackClient = (channel: string, socketInstance: SocketInstance): string => {
+  const socketId = uuidv4();
+  sockets[socketId] = socketInstance;
+
+  if(channel in channels) {
+    channels[channel].sockets.push(socketId)
+  } else {
+    channels[channel] = {sockets: [socketId]};
+  }
+
+  return socketId;
+}
+
+export const sendToChannel = (channel: string, value: EventValue): void => {
+  const strData = JSON.stringify(value);
+  if(!channels[channel]) {
+    console.debug(`channel ${channel} is unknown, skipping`);
+    return;
+  }
+  channels[channel].sockets.forEach(socketId => {
+    const socketInstance: SocketInstance = sockets[socketId];
+    if(!socketInstance) return cleanChannel(channel);
+    try {
+      socketInstance.ws.send(strData);
+    } catch(err) {
+      console.debug('Error sending to socket', err);
```

Comment:
This can happen for a number of reasons, including a disconnected socket. I believe this was a `console.error` and was downgraded to a debug due to noise.